torch/export/_trace.py

```diff
@@ -105,6 +105,29 @@ def _ignore_backend_decomps():
         torch.backends.nnpack.set_flags(*orig_nnpack_flag)
 
 
+@contextmanager
+def _patch_dropout_decomps():
+    """
+    In export, we always want to decompose dropout_
+    to functional dropout. The very correct way to do is
+    we extend functionalization to automatically functionalize
+    CompositeImplicitAutograd ops using their functional variant.
+    But this is significantly lot more work to do. Since, it seems
```

Comment:
Hmm, what's the argument for not doing this the normal way? Just thinking:

(1) If we want to advertise pre-dispatch functionalization as truly functional, that means we'll need to do the work of going through the CompositeImplicitAutograd ops with incorrect schemas. Which in theory we can do by extending `SchemaCheckMode` a bit so we can put it on the pre-dispatch mode stack, and then relying on OpInfo coverage to get example inputs to all of our CompositeImplicitAutograd ops. This... might actually work out ok? OpInfos are mostly written for ops in the torch.* namespace, and most CompositeImplicitAutograd end up getting called by a torch.* op. Happy to brainstorm some more.

(2) If we want to enable a small number of mutable CompositeImplicitAutograd ops in the short term, it would probably be better to add a registry to `FunctionalTensorMode`, instead patching in a temporary set of decompositions.

Reply:
For 2) How would we actually do that inside FunctionalTensorMode? Are you saying we override  the implementation of func.decompose()? Would it be better if we just write a functionalization rule for dropout_ by hand?

Reply:
> Would it be better if we just write a functionalization rule for dropout_ by hand?

Pretty much this. We do something similar for FakeTensor (it has a registry of FakeTensor rules manually written, e.g. [here](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/fake_impls.py#L261)).

The most consistent thing would be to hand-write the rule for it in C++ (so we have less drift between python and C++ functionalization). If we think it won't be very long-lived though (since we want functionalization to work properly for all compositeImplicit ops), then putting it in python seems fine, similar to how FakeTensor's registry works.

LMK if you want to talk about it some more!